engine/input.asm

```diff
@@ -4,82 +4,131 @@
 
 nop
 
+.struct InputState
+    index  db
+    start  db
+    select db
+    upbtn  db
+    dnbtn  db
+    lftbtn db
+    rhtbtn db
+.endst
+
 .struct Input
-    frame_counter dw
-    input1l_q instanceof Queue
-    input1h_q instanceof Queue
+    inputstate instanceof InputState
+    enabled db
 .endst
 
 .enum $00
     input instanceof Input
 .ende
 
-;Defining and init Queue here does not seem to be pushing/pop
-;Defining and init in Main seems to work for push, but it keeps pushing over memory
-;Queue_Define("INPUT1L_Q", $0090, #$0002, #$0004)
-;Queue_Define("INPUT1H_Q", $00A0, #$0002, #$0004)
-
 Input_Init:
-    pha
-    phy
-    phx
-
-    ; Allocate Queue + 4 bytes for the queue
-    lda #(_sizeof_Queue)
-    adc #$4
-    jsr Malloc_Bytes
-
-    ; X has the start address
-    ; Y has the end address
-    jsr Queue_Init
-
-    txy         ; Y is now the start address
-    lda 1, S    ; Grab the this pointer again
-    tax         ; X is now the proper offset
-    tya         ; Put the start address back into the accumulator
-    sta input.input1l_q, X ; Store the start address into the input struct
-
-    ; Allocate Queue + 4 bytes for the second queue
-    lda #(_sizeof_Queue)
-    adc #$4
-    jsr Malloc_Bytes
-
-    ; X has the start address
-    ; Y has the end address
-    jsr Queue_Init
-
-    txy         ; Y is now the start address
-    lda 1, S    ; Grab the this pointer again
-    tax         ; X is now the proper offset
-    tya         ; Put the start address back into the accumulator
-    sta input.input1h_q, X ; Store the start address into the input struct
-
-    lda #$0000
-    sta input.frame_counter, X
-
-    plx
-    ply
-    pla
+    A8
+    stz input.inputstate.upbtn, X
+    stz input.inputstate.dnbtn, X
+    stz input.inputstate.rhtbtn, X
+    stz input.inputstate.lftbtn, X
+    A16
     rts
 
 Input_Frame:
     rts
 
-; X is "this" pointer
 Input_VBlank:
     pha
-
-    inc input.frame_counter, X
-
-    lda JOY1L
-    call_ptr(Queue_Push, input.input1l_q)
-    call_ptr(Queue_Pop, input.input1l_q)
-
-    lda JOY1H
-    call_ptr(Queue_Push, input.input1h_q)
-    call_ptr(Queue_Pop, input.input1h_q)
-
+    jsr Input_UpButton
+    jsr Input_DnButton
+    jsr Input_LftButton
+    jsr Input_RhtButton
     pla
     rts
 
+Input_DnButton:
+    pha
+    @CheckDnButton:
+        lda JOY1L                          ; check whether the Dn button was pressed this frame...
+        cmp #DNBTN
```

Comment:
You could turn this into a [16-bit `bit #DNBTN`](http://6502.org/tutorials/65c816opcodes.html#6.1.2.2) so that you are testing whether a bit is set for your controller reads. Your branching remains the same. In your code now, you can only have down pressed and nothing else. It would probably simplify your code a bit too.

```suggestion
        bit #DNBTN
```